File: sdks/php/tests/Api/EntityApiTest.php
<?php

declare(strict_types=1);

namespace Atap\Sdk\Tests\Api;

use Atap\Sdk\Model\Entity;
use Atap\Sdk\Model\KeyVersion;
use Atap\Sdk\Tests\TestCase;

class EntityApiTest extends TestCase
{
    public function testRegister(): void
    {
        $client = $this->createUnauthClient([
            $this->jsonResponse(201, [
                'id' => 'entity123',
                'type' => 'agent',
                'did' => 'did:web:localhost%3A8080:agent:entity123',
                'name' => 'Test Agent',
                'key_id' => 'key_123',
                'public_key' => 'pubkey_b64',
                'client_secret' => 'atap_secret123',
            ]),
        ]);

        $entity = $client->entities()->register('agent', 'Test Agent');

        $this->assertInstanceOf(Entity::class, $entity);
        $this->assertSame('entity123', $entity->id);
        $this->assertSame('agent', $entity->type);
        $this->assertSame('Test Agent', $entity->name);
        $this->assertSame('atap_secret123', $entity->clientSecret);
    }

    public function testRegisterWithPublicKey(): void
    {
        $client = $this->createUnauthClient([
            $this->jsonResponse(201, [
                'id' => 'entity456',
                'type' => 'machine',
                'did' => 'did:web:localhost%3A8080:machine:entity456',
                'public_key' => 'my_pubkey',
            ]),
        ]);

        $entity = $client->entities()->register('machine', '', 'my_pubkey');
        $this->assertSame('entity456', $entity->id);
        $this->assertSame('machine', $entity->type);
    }

    public function testGet(): void
    {
        $client = $this->createUnauthClient([
            $this->jsonResponse(200, [
                'id' => 'entity789',
                'type' => 'human',
                'did' => 'did:web:localhost%3A8080:human:entity789',
                'trust_level' => 2,
            ]),
        ]);

        $entity = $client->entities()->get('entity789');
        $this->assertSame('entity789', $entity->id);
        $this->assertSame('human', $entity->type);
        $this->assertSame(2, $entity->trustLevel);
    }

    public function testDelete(): void
    {
        $client = $this->createAuthedClient([
            $this->noContentResponse(),
        ]);

        // Should not throw
        $client->entities()->delete('entity123');
        $this->assertTrue(true);
    }

    public function testRotateKey(): void
    {
        $client = $this->createAuthedClient([
            $this->jsonResponse(200, [
                'id' => 'key_new_123',
                'entity_id' => 'entity123',
                'key_index' => 2,
                'valid_from' => '2024-01-01T00:00:00Z',
                'created_at' => '2024-01-01T00:00:00Z',
            ]),
        ]);

        $keyVersion = $client->entities()->rotateKey('entity123', 'new_pubkey_b64');
        $this->assertInstanceOf(KeyVersion::class, $keyVersion);
        $this->assertSame('key_new_123', $keyVersion->id);
        $this->assertSame(2, $keyVersion->keyIndex);
    }

    public function testRegisterWithPrincipalDid(): void
    {
        $client = $this->createUnauthClient([
            $this->jsonResponse(201, [
                'id' => 'agent_sub',
                'type' => 'agent',
                'principal_did' => 'did:web:localhost%3A8080:human:parent',
            ]),
        ]);

        $entity = $client->entities()->register(
            'agent',
            '',
            null,
            'did:web:localhost%3A8080:human:parent',
        );
        $this->assertSame('did:web:localhost%3A8080:human:parent', $entity->principalDid);
    }
}

File: sdks/java/src/test/java/dev/atap/sdk/api/EntityApiTest.java
package dev.atap.sdk.api;

import com.github.tomakehurst.wiremock.WireMockServer;
import com.github.tomakehurst.wiremock.core.WireMockConfiguration;
import dev.atap.sdk.ATAPClient;
import dev.atap.sdk.TestFixtures;
import dev.atap.sdk.model.Entity;
import dev.atap.sdk.model.KeyVersion;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static com.github.tomakehurst.wiremock.client.WireMock.*;
import static org.junit.jupiter.api.Assertions.*;

class EntityApiTest {

    private WireMockServer wireMock;
    private ATAPClient client;

    @BeforeEach
    void setUp() {
        wireMock = new WireMockServer(WireMockConfiguration.wireMockConfig().dynamicPort());
        wireMock.start();

        // Stub token endpoint for authenticated requests
        wireMock.stubFor(post(urlEqualTo("/v1/oauth/token"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(TestFixtures.tokenResponseJson())));

        client = ATAPClient.builder()
                .baseUrl("http://localhost:" + wireMock.port())
                .did(TestFixtures.TEST_DID)
                .keyPair(TestFixtures.getTestKeyPair())
                .clientSecret(TestFixtures.TEST_CLIENT_SECRET)
                .platformDomain(TestFixtures.TEST_PLATFORM_DOMAIN)
                .build();
    }

    @AfterEach
    void tearDown() {
        wireMock.stop();
        client.close();
    }

    @Test
    void testRegister() {
        wireMock.stubFor(post(urlEqualTo("/v1/entities"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(TestFixtures.entityResponseJson())));

        Entity entity = client.entities.register("agent", "Test Agent", null, null);
        assertNotNull(entity);
        assertEquals("agent", entity.getType());
        assertEquals("Test Agent", entity.getName());
        assertEquals(TestFixtures.TEST_ENTITY_ID, entity.getId());
    }

    @Test
    void testRegisterMinimal() {
        wireMock.stubFor(post(urlEqualTo("/v1/entities"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(TestFixtures.entityResponseJson())));

        Entity entity = client.entities.register("agent");
        assertNotNull(entity);
        assertEquals("agent", entity.getType());
    }

    @Test
    void testGet() {
        wireMock.stubFor(get(urlEqualTo("/v1/entities/" + TestFixtures.TEST_ENTITY_ID))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(TestFixtures.entityResponseJson())));

        Entity entity = client.entities.get(TestFixtures.TEST_ENTITY_ID);
        assertNotNull(entity);
        assertEquals(TestFixtures.TEST_ENTITY_ID, entity.getId());
        assertEquals(TestFixtures.TEST_DID, entity.getDid());
        assertEquals(1, entity.getTrustLevel());
    }

    @Test
    void testDelete() {
        wireMock.stubFor(delete(urlEqualTo("/v1/entities/" + TestFixtures.TEST_ENTITY_ID))
                .willReturn(aResponse().withStatus(204)));

        assertDoesNotThrow(() -> client.entities.delete(TestFixtures.TEST_ENTITY_ID));
    }

    @Test
    void testRotateKey() {
        String keyVersionJson = "{\"id\":\"key_v2\","
                + "\"entity_id\":\"" + TestFixtures.TEST_ENTITY_ID + "\","
                + "\"key_index\":2,"
                + "\"valid_from\":\"2025-01-02T00:00:00Z\","
                + "\"created_at\":\"2025-01-02T00:00:00Z\"}";

        wireMock.stubFor(post(urlEqualTo("/v1/entities/" + TestFixtures.TEST_ENTITY_ID + "/keys/rotate"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(keyVersionJson)));

        KeyVersion kv = client.entities.rotateKey(TestFixtures.TEST_ENTITY_ID, "newpubkey==");
        assertNotNull(kv);
        assertEquals("key_v2", kv.getId());
        assertEquals(2, kv.getKeyIndex());
        assertEquals(TestFixtures.TEST_ENTITY_ID, kv.getEntityId());
    }

    @Test
    void testRegisterWithPublicKey() {
        String responseWithSecret = "{\"id\":\"newid\","
                + "\"type\":\"agent\","
                + "\"did\":\"did:web:example:agent:newid\","
                + "\"name\":\"My Agent\","
                + "\"key_id\":\"key_test\","
                + "\"public_key\":\"testkey==\","
                + "\"client_secret\":\"atap_secretvalue\","
                + "\"trust_level\":0,"
                + "\"created_at\":\"2025-01-01T00:00:00Z\","
                + "\"updated_at\":\"2025-01-01T00:00:00Z\"}";

        wireMock.stubFor(post(urlEqualTo("/v1/entities"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(responseWithSecret)));

        Entity entity = client.entities.register("agent", "My Agent", "testkey==", null);
        assertNotNull(entity);
        assertEquals("atap_secretvalue", entity.getClientSecret());
    }
}
